xdsl/dialects/comb.py

```diff
@@ -0,0 +1,245 @@
+from typing import Annotated
+
+from xdsl.dialects.builtin import IndexType, IntegerAttr, IntegerType, UnitAttr, i32
+from xdsl.ir import Dialect, OpResult
+from xdsl.irdl import (
+    ConstraintVar,
+    IRDLOperation,
+    Operand,
+    VarOperand,
+    attr_def,
+    irdl_op_definition,
+    operand_def,
+    opt_attr_def,
+    result_def,
+    var_operand_def,
+)
+
+
+class BinCombOp(IRDLOperation):
+    """
+    A binary comb operation. It has two operands and one
+    result, all of the same integer type.
+    """
+
+    T = Annotated[IntegerType, ConstraintVar("T")]
+
+    lhs: Operand = operand_def(T)
+    rhs: Operand = operand_def(T)
+    result: OpResult = result_def(T)
+
+    two_state: UnitAttr | None = opt_attr_def(UnitAttr)
+
+
+class VariadicCombOp(IRDLOperation):
```

Comment:
```suggestion
class VariadicCombOperation(IRDLOperation, abc.ABC):
```

We've been annotating abstract base operation classes explicitly with `abc.ABC`, and appending `Operation` as opposed to `Op` as a convention. It's not documented anywhere and I'm not sure it's 100% consistent across all dialects but I find it pretty useful as a visual guide, could you please add it here also? This is more of a nit, so feel free to ignore.

Reply:
Can you also please add an `__init__` to the two base classes?

Reply:
We were thinking of adding the `__init__` after in this case?
So we can have more reasonable PRs ;)
Also, same things for the parser/printer

Reply:
Working on finishing the __init__ and parser/printer this morning :)

Reply:
I would recommend merging operations one by one but in finished form rather than leaving bits for later but I guess that's a subjective call. We merged the riscv dialect almost instruction by instruction and I thought that went quite well.

Reply:
I'm fine with both, so fine for me!

Reply:
Actually I'm not sure.
Since everything is already implemented now, can we just merge this and then add the __init__? The __init__ should not be a breaking change, and adding __init__ here would just make the PR so much bigger (since we would add the tests as well?)

Reply:
Sure that's not a problem

Reply:
I was wondering about the "merging operations one by one" and I thought of suggesting it, but since now everything is in place, it rather impedes progress.

I've tried to look in git history how has this been handled in other dialects, but I've mostly stumbled in PRs that moved files around, so not sure what might be a good practice.